server/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReference.java

```diff
@@ -182,60 +181,85 @@ private static void advance(final BytesRef ref, final int length) {
     }
 
     /**
-     * Instead of adding the complexity of {@link InputStream#reset()} etc to the actual impl
-     * this wrapper builds it on top of the BytesReferenceStreamInput which is much simpler
-     * that way.
+     * Mark-supporting {@link StreamInput} over all the bytes in a {@link BytesReference}.
      */
     private static final class MarkSupportingStreamInputWrapper extends StreamInput {
-        // can't use FilterStreamInput it needs to reset the delegate
         private final BytesReference reference;
-        private BytesReferenceStreamInput input;
+        private int offset = 0;
         private int mark = 0;
 
-        private MarkSupportingStreamInputWrapper(BytesReference reference) throws IOException {
+        private MarkSupportingStreamInputWrapper(BytesReference reference) {
             this.reference = reference;
-            this.input = new BytesReferenceStreamInput(reference.iterator(), reference.length());
         }
 
         @Override
         public byte readByte() throws IOException {
-            return input.readByte();
+            if (offset >= reference.length()) {
+                throw new EOFException();
+            }
+            return reference.get(offset++);
         }
 
         @Override
-        public void readBytes(byte[] b, int offset, int len) throws IOException {
-            input.readBytes(b, offset, len);
+        public int readInt() {
+            final int res = reference.getInt(offset);
+            offset += Integer.BYTES;
+            return res;
+        }
+
+        @Override
+        public void readBytes(byte[] b, int offset, int len) {
+            final int read = read(b, offset, len);
+            if (read != len) {
+                throw new IndexOutOfBoundsException();
```

Comment:
Maybe `EOFException` would be more appropriate? Hitting this means we ran out of bytes AFAICT.

Reply:
Yea I'd agree but we have a bunch of tests that verify we run into `IndexOutOfBoundsException` and I'm a little hesitant to change behavior from a `RuntimeException` to an `IOException` here when we have a few spots in upstream code that catch `IOException`s and where we could then accidentally suppress de-serialization bugs?

Reply:
Oh I see this is copied from `BytesReferenceStreamInput`. Hmm. This is all a bit weird, `readByte()` throws an `EOFException` if it runs out of data, as does `ensureCanReadBytes`. Also today all the structured `readX()` methods like `readInt()` use repeated `readByte()` calls so they yield `EOFException`s too. `BytesReferenceStreamInput#read` might throw an `EOFException` or an `ArrayIndexOutOfBoundsException` depending on how exactly it runs out of space. What a mess.